// @applitools/core-base/dist/heartbeat.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeHeartbeat = void 0;
const utils = __importStar(require("@applitools/utils"));
let processId = utils.general.guid();
let isRunning = false;
let intervalId;
let refCount = 0;
function makeHeartbeat({ requests, logger: mainLogger }) {
    const logger = mainLogger.extend(mainLogger, { tags: [`heartbeat-base-${utils.general.shortid()}`] });
    function stopHeartbeats() {
        logger.log('Stopping heartbeats');
        if (intervalId) {
            clearInterval(intervalId);
            intervalId = undefined;
            isRunning = false;
            processId = utils.general.guid();
            logger.log('Generated new processId after expiration:', processId);
        }
    }
    async function startHeartbeats(settings) {
        // send the first heartbeat right away
        await requests.sendHeartbeat({ settings: { ...settings, processId }, logger }).catch(error => {
            logger.error('Failed to send heartbeat:', error);
        });
        intervalId = setInterval(async () => {
            try {
                await requests.sendHeartbeat({ settings: { ...settings, processId }, logger });
            }
            catch (error) {
                if (error.message === 'This process is expired. All its sessions were abandoned') {
                    logger.log('Heartbeat process expired on server');
                    // Process expired - stop and reset so next acquire will generate new processId
                    stopHeartbeats();
                }
            }
        }, settings.interval);
        intervalId.unref();
    }
    function acquire(settings) {
        refCount++;
        logger.log('Heartbeat acquired, refCount:', refCount);
        if (!isRunning) {
            // Use provided processId only if it matches our current processId
            // This ensures we don't reuse an expired processId
            if (settings.processId && settings.processId === processId) {
                logger.log('Starting heartbeats with provided processId:', processId);
            }
            else if (!settings.processId || settings.processId !== processId) {
                // Either no processId provided, or it doesn't match current one
                // Generate new one (current processId was already reset if expired)
                if (settings.processId && settings.processId !== processId) {
                    logger.log('Provided processId differs from current, using current processId:', processId);
                }
                else {
                    processId = utils.general.guid();
                    logger.log('Starting heartbeats with new processId:', processId);
                }
            }
            isRunning = true;
            // Start heartbeat in background
            void startHeartbeats(settings).catch(error => {
                logger.error('Failed to start heartbeats:', error);
            });
        }
        return processId;
    }
    function release() {
        if (refCount > 0) {
            refCount--;
            logger.log('Heartbeat released, refCount:', refCount);
            if (refCount === 0) {
                // Last session closed - stop heartbeat
                stopHeartbeats();
            }
        }
    }
    return {
        get processId() {
            return processId;
        },
        acquire,
        release,
    };
}
exports.makeHeartbeat = makeHeartbeat;
